knoten/sensor_utils.py

```diff
@@ -0,0 +1,296 @@
+                        illum_pos.z - ground_pt.z)
+  
+    return np.rad2deg(utils.sep_angle(vec_a, vec_b))
+
+def emission_angle(image_pt, sensor):
+    """
+    Computes and returns emission angle, in degrees, for a given image point.
+
+    Emission Angle: The angle between the surface normal vector at the
+    intersection point and the vector from the intersection point to the
+    observer (usually the spacecraft). The emission angle varies from 0 degrees
+    when the observer is viewing the sub-spacecraft point (nadir viewing) to 90
+    degrees when the intercept is tangent to the surface of the target body.
+    Thus, higher values of emission angle indicate more oblique viewing of the
+    target.
+
+    Parameters
+    ----------
+    image_pt : tuple
+               Pair of x, y (sample, line) coordinates in pixel space
+
+    sensor : object
+             A CSM compliant sensor model object
+
+    Returns
+    -------
+     : np.ndarray
+            emission angle in degrees
+    """
+    if not isinstance(image_pt, csmapi.ImageCoord):
+        image_pt = csmapi.ImageCoord(*image_pt)
+
+    sensor_state = csm.get_state(sensor, image_pt)
+    ground_pt = csm.generate_ground_point(0.0, image_pt, sensor)
+
+    normal = csm.get_surface_normal(sensor, ground_pt)
+
+    sensor_diff = utils.Point(sensor_state["sensorPos"].x - ground_pt.x, 
+                              sensor_state["sensorPos"].y - ground_pt.y,
+                              sensor_state["sensorPos"].z - ground_pt.z)
+    
+    return np.rad2deg(utils.sep_angle(normal, sensor_diff))
+
+def slant_distance(image_pt, sensor):
+    """
+    Computes the slant distance from the sensor to the ground point.
+
+    Parameters
+    ----------
+    image_pt : tuple
+               Pair of x, y (sample, line) coordinates in pixel space
+
+    sensor : object
+             A CSM compliant sensor model object
+
+    Returns
+    -------
+     : np.ndarray 
+        slant distance in meters
+    """
+    if not isinstance(image_pt, csmapi.ImageCoord):
+        image_pt = csmapi.ImageCoord(*image_pt)
+
+    sensor_state = csm.get_state(sensor, image_pt)
+    ground_pt = csm.generate_ground_point(0.0, image_pt, sensor)
+
+    return utils.distance(sensor_state["sensorPos"], ground_pt)
+
+def target_center_distance(image_pt, sensor):
+    """
+    Calculates and returns the distance from the spacecraft to the target center.
+
+    Parameters
+    ----------
+    image_pt : tuple
+               Pair of x, y (sample, line) coordinates in pixel space
+
+    sensor : object
+             A CSM compliant sensor model object
+
+    Returns
+    -------
+     : np.ndarray 
+        target center distance in meters
+    """
+    if not isinstance(image_pt, csmapi.ImageCoord):
+        image_pt = csmapi.ImageCoord(*image_pt)
+
+    sensor_state = csm.get_state(sensor, image_pt)
+    return utils.distance(sensor_state["sensorPos"], utils.Point(0,0,0))
+
```

Comment:
Is there a particular reason we avoided the illuminationDistance function? I know it won't "work" but it would be nice to provide the method signature as it was in the original library

It seems like other illunmination functions were also omitted, could we include them for the same reason as to above?

Reply:
This is why the illumination functions were excluded: https://github.com/DOI-USGS/ISIS3/blob/21308692275a243ff58aeefce996cb3c0d8c357b/isis/src/base/objs/CSMCamera/CSMCamera.cpp#L1209-L1233. They all are NULL in campt due to CSM not supporting them. Sounds like using `getIlluminationDirection` gets us "close" to the correct solar position but it's not accurate. I can add them back if we are okay with the results not matching ISIS.

Reply:
I would like to see them ported just so they don't get lost in translation but am open to other ideas!